### PR TITLE
Geselowitz ECG

### PR DESCRIPTION
**Pull request by AbdAlazezAhmed**
For #71 and #72

### CODE REVIEW

File: src/solver/ecg.jl [hunk 1]
```diff
@@ -59,6 +40,147 @@ function Plonsey1964ECGGaussCache(problem::SplitProblem{<:TransientHeatProblem},
...
+
+    ch = ConstraintHandler(dh)
+    Ferrite.add!(ch, Dirichlet(:Z, Set([zero_vertex]), (x, t) -> 0.0))
+    close!(ch);
+    apply!(op.A, f, ch)
+
+    for (i, electrode_pair) in enumerate(electrode_pairs)
+        _add_electrodes(f, dh, "lead_field_$(i)", electrodes[electrode_pair[1]], electrodes[electrode_pair[2]])
+        ml = ruge_stuben(op.A)
+        p = aspreconditioner(ml)
```
Comment (termi-official) [inline src/solver/ecg.jl:116]: You can hoist this out of the loop.

File: src/solver/ecg.jl [hunk 2]
```diff
@@ -95,3 +217,95 @@ function _evaluate_ecg_plonsey_gauss(κ∇φₘ, coords::AbstractVector{Vec{dim,
     end
     return φₑ_local
 end
+
+"""
+    evaluate_ecg(method::Geselowitz1989ECGLeadCache, φₘ::Vector, κ::SpectralTensorCoefficient)
+
+Compute the lead potential between lead pairs by evaluating:
```
Comment (termi-official) [inline src/solver/ecg.jl:224]:
```suggestion
Compute the potential between lead pairs by evaluating:
```

File: src/solver/ecg.jl [hunk 3]
```diff
@@ -59,6 +40,147 @@ function Plonsey1964ECGGaussCache(problem::SplitProblem{<:TransientHeatProblem},
...
+    local_positive_dof = Ferrite.vertexdof_indices(Ferrite.getfieldinterpolation(dh.subdofhandlers[1], :Z))[positive_vertex[2]][1]::Int
+    global_positive_dof = celldofs(dh, positive_vertex[1])[local_positive_dof]::Int
+    f[global_positive_dof] = -1
+    local_negative_dof = Ferrite.vertexdof_indices(Ferrite.getfieldinterpolation(dh.subdofhandlers[1], :Z))[negative_vertex[2]][1]::Int
+    global_negative_dof = celldofs(dh, negative_vertex[1])[local_negative_dof]::Int
+    f[global_negative_dof] = 1
+end
+
+"""
+    TODORENAMEECGPoissonReconstructionCache(mesh, κ, qr_collection, ip_collection, zero_vertex::VertexIndex)
```
Comment (termi-official) [inline src/solver/ecg.jl:140]:
```suggestion
    ECGPoissonReconstructionCache(mesh, κ, qr_collection, ip_collection, zero_vertex::VertexIndex)
```

I think this was first proposed by https://ieeexplore.ieee.org/document/4015619 . We should also mention https://ieeexplore.ieee.org/abstract/document/5759732 as a method to improve the precision further. The FIMH paper is another analysis trying to explain what exactly breaks down.

**Review by termi-official — COMMENTED** [review]
I assume further progress is blocked until we have subdomains or intergrid transfer operators merged?

**Comment by termi-official** [comment]
You should be able to revert the hotfixes in src/solver and src/modeling and merge master.

There is now also a nodal transfer operator (not tested) which can be used to transfer the solution between heart and torso.

File: test/integration/test_ecg.jl
Comment (termi-official) [inline test/integration/test_ecg.jl:1]: Can we add idempotence checks for the ecg update?

File: test/integration/test_ecg.jl [hunk 4]
```diff
@@ -66,10 +79,44 @@
...
+            end
+        end
+
+        @testset "Idempotence" begin
+            u .= [randn() for _ in u]
+            u_g .= [randn() for _ in u_g]
+
+            @testset "Plonsey1964" begin
+                Thunderbolt.update_ecg!(plonsey_ecg, u)
+                val_1_init = Thunderbolt.evaluate_ecg.(Ref(plonsey_ecg), electrodes, 1.0)
```
Comment (termi-official) [inline test/integration/test_ecg.jl:97]: Why do we need to Ref here?

**Review by termi-official — COMMENTED** [review]
LGTM otherwise.

File: test/integration/test_ecg.jl (hunk 4 above)
Comment (AbdAlazezAhmed) [reply test/integration/test_ecg.jl:97]: because the broadcasting tries to iterate over the struct iiuc

**Review by termi-official — APPROVED** [review]

**Pull request merged**